scvi/external/cell2location/_module.py

```diff
@@ -0,0 +1,336 @@
+        
+        self.cell_state_mat = cell_state_mat
+        self.register_buffer("cell_state", torch.tensor(cell_state_mat.T))
+        
+        self.guide = AutoNormal(self.model, init_loc_fn=init_to_mean, 
+                                create_plates=self.create_plates)
+        # replace exp transform with softplus https://github.com/pyro-ppl/numpyro/issues/855
+        #self.guide = ExpToSoftplusHandler(self.guide)
+
+    @staticmethod
+    def _get_fn_args_from_batch(tensor_dict):
+        x_data = tensor_dict[_CONSTANTS.X_KEY]
+        ind_x = tensor_dict["ind_x"].long().squeeze()
+        obs2sample = tensor_dict["obs2sample"]
+        return (x_data, ind_x, obs2sample), {}
+    
+    @staticmethod
+    def _get_fn_args_for_predictive(adata):
+        x_data = torch.tensor(adata.layers['counts'][0:100,:].toarray().astype('float32'))
+        ind_x = torch.tensor(adata.obs['_indices'].values.astype('int64').squeeze()).long()
+        obs2sample = torch.tensor(adata.obsm['obs2sample_obsm'].astype('float32'))
+        return (x_data, ind_x, obs2sample), {}
+    
+    
+    def create_plates(self, x_data, idx, obs2sample):
+        return [pyro.plate("obs_axis", self.n_obs, dim=-2, 
+                           subsample_size=self.batch_size, 
+                           subsample=idx),
+                pyro.plate("var_axis", self.n_var, dim=-1),
+                pyro.plate("factor_axis", self.n_fact, dim=-1),
+                pyro.plate("combination_axis", self.n_comb, dim=-3),
+                pyro.plate("experim_axis", self.n_exper, dim=-2)]
+
+    def model(self, x_data, idx, obs2sample):
+        # register module with Pyro
+        pyro.module("cell2location", self)
+        
+        obs_axis, var_axis, factor_axis, combination_axis, experim_axis = self.create_plates(x_data, idx, obs2sample)
+        
+        # =====================Gene expression level scaling m_g======================= #
+        # Explains difference in sensitivity for each gene between single cell and spatial technology
+        # compute hyperparameters from mean and sd
+        shape = self.m_g_gene_level_prior['mean'] ** 2 / self.m_g_gene_level_prior['sd'] ** 2
+        rate = self.m_g_gene_level_prior['mean'] / self.m_g_gene_level_prior['sd'] ** 2
+        shape_var = shape / self.m_g_gene_level_prior['mean_var_ratio']
+        rate_var = rate / self.m_g_gene_level_prior['mean_var_ratio']
+        
+        m_g_alpha_hyp = pyro.sample('m_g_alpha_hyp',
+                                                Gamma(mu=shape,
+                                                      sigma=shape_var,
+                                                      shape=(1, 1)))
+
+        m_g_beta_hyp = pyro.sample('m_g_beta_hyp',
+                                               Gamma(mu=rate,
+                                                     sigma=rate_var,
+                                                     shape=(1, 1)))
```

Comment:
I believe another way to make this device agnostic is to leverage the device of `x_data`. So you'd make a torch tensor and use the `dev = x_data.device`. You can also give `pyro.sample` a lambda function as input like `pyro.sample("m_g_beta_hyp", lambda self: Gamma(mu=self.rate, sigma=self.rate_var)` where rate and rate_var would be registered buffers.

Reply:
Is the lambda function approach necessary if `self.rate, sigma=self.rate_var` are registered buffers?